File: Neetcode Questions/Linked List/reorder-list.cpp

// https://leetcode.com/problems/reorder-list/description/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if(!head) return;

        vector<int> arr;
        ListNode*temp = head;
        while(temp){
            arr.push_back(temp->val);
            temp = temp->next;
        }

        temp = head;
        int i = 0, j = arr.size()-1;
        for( ; i < j; i++, j--){
            temp->val = arr[i];
            temp = temp->next;
            temp->val = arr[j];
            temp = temp->next;
        }

        if(i == j) temp->val = arr[i];
        return;
    }
};

File: Neetcode-150-js/Linked List/reorder-list.js

// https://leetcode.com/problems/reorder-list/description/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    if(!head || !head.next) return head;

    const arr = [];
    let temp = head;
    while(temp){
        arr.push(temp.val);
        temp = temp.next;
    }
    
    temp = head;
    let i = 0, j = arr.length-1;
    while(i <= j){
        temp.val = arr[i++];
        temp = temp.next;
        if(temp){
            temp.val = arr[j--];
            temp = temp.next;
        }
    }
    return head;
};